README.md
# 项目介绍

有时候，我们在外想要访问家里或公司主机的资料，要么由于主机处于内网下，要么就是是运营商随机分配的一个公网IP，都很难直接连上主机获取资料。

那么，有什么办法可以解决这一难题？答案就是**内网穿透**。当内网中的主机没有静态IP地址要被外网稳定访问时可以使用内网穿透。


## 功能列表

- 基于 TCP 协议
- 支持多端口穿透
- 支持断线重连
- 支持指定访问端口
- 支持身份认证，引入基于有效期的身份失效机制
- 单服务支持多个连接

## 工作原理

![chuantou_architecture](doc/chuantou_architecture.png)

通常我们需要访问一个服务，只需要直接通过地址连接就可以了，但是如果服务处在一个与访问者隔离的内网环境里，我们就不能访问了。
不过也不是没有办法，如果内网服务可以访问外网的话，我们就可以找到突破点。

### 实现步骤

**服务端部分**
1. 启动`tunnel_bridge`， 监听`tunnel_client`的连接。
2. 启动`tunnel_server`， 监听`用户`的访问请求。

**客户端部分**
1. 启动`tunnel_client`， 向`tunnel_bridge`拨号，建立连接。
2. 如果`用户`发起请求，则向`内网服务`拨号，建立连接。
3. 处理连接双方的通讯，实现穿透。

## 如何部署实现内网穿透

**服务端** 需要一台具有公网 IP 的服务器（比如阿里云），配置不限，将服务端部署上去。  
**客户端** 部署在内网服务器，与需要穿透的服务（比如 mysql）在同一个局域网。

> 需要注意的是，实现穿透后，就相当于将服务暴露在互联网，会面临被攻击的风险。

## 如何编译

目前改用了 go.mod 方式管理依赖，在国内使用最好配置代理，否则会出现无法下载的情况。

可以通过设置环境变量实现，idea 可以直接设置代理
```
export GOPROXY=https://goproxy.io
```

```shell script
# linux
$ go build -o chuantou main.go

# windows
$ go build -o chuantou.exe main.go
```


## 生成带失效时间的“短期 key”

> 注意：此 key 仅用于客户端，服务端用的还是原 key

```shell script
# 假如希望生成到期日期为 2019-12-31 的 key，使用命令：

$ chuantou -generate <key> [expired-time]

# key                原 key ， 必填参数
# expired-time       到期日期，如：2019-12-31，未输入此参数的话，生成有效期为一个月的 key

```

## 启动方式

支持两种方式启动:

- 纯命令行启动（不需要配置文件）
- 配置文件启动（需要配置文件）**【推荐】**

### 命令行启动

```shell script
# 启动服务端

$ chuantou -server <key> <port>

# 注释
# key                 建议长度 6-16 个字符，用于身份校验
# port                服务端端口，不要使用保留端口，必填
# access-port-range   访问端口范围，必填，如 10000-20000

```

```shell script
# 启动客户端

$ chuantou -client <key> <server:port> <local:port:mapping> [tunnel-count]

# 注释
# key                与服务端保持一致
# server:port        服务端地址，格式如：45.32.78.129:6666
# local:port:mapping 被代理服务地址及访问端口（访问端口省略时表示不进行端口转换），多个以逗号隔开，比如：127.0.0.1:3389,127.0.0.1:3306:13306
# tunnel-count       隧道条数，默认为1，范围[1-5]
```

### 配置文件启动

**启动前准备**  
将`config_demo.ini`重命名为`config.ini`

**服务端配置**
```ini
# 服务端配置
[server]
# 代理端口
port = 6666
# Key 建议长度 6-16 个字符，用于身份校验
key = winshu
```

**客户端配置**
```ini
# 客户端配置
[client]
# 与服务端保持一致
key = winshu
# 服务端地址，格式 ip:port
server-host = 127.0.0.1:6666
# 内网被代理服务地址及访问端口(多个用逗号隔开)，格式如 192.168.1.100:3389:13389
# 内网IP:内网端口:访问端口
local-host-mapping = ["127.0.0.1:3306:13307","127.0.0.1:3389:13389"]
# 隧道条数，默认为1，范围[1-5]
tunnel-count = 1
```

**启动命令**
```shell script
# 启动服务端
$ chuantou -server

# 启动客户端
$ chuantou -client
```



doc/CHANGELOG.md
# 版本变动日志

## Version 1.0.0

- 实现内网穿透
- 支持多端口穿透
- 服务端生成代理端口时回写给客户端，解决随机端口的问题

## Version 1.0.1
- 增加代理接入身份校验，解决服务端能被任何人使用的问题
- 支持客户端指定代理端口模式

## Version 1.0.2
- 使用延迟创建内网服务连接的方式，解决不停连接内网服务的问题
- 解决内网服务断开，访问服务仍未关闭的问题
- 去除随机端口访问模式
- 简化通讯协议

## Version 1.0.3
- 单服务支持多个连接，服务更稳定
- 通讯协议加入原端口

## Version 1.0.4
- 解决隧道桥中断，连接没有及时处理的问题
- 解决断线不能真正重连的问题

## Version 1.0.5
- 引入基于有效期的身份失效机制
- 增加版本校验机制，避免客户端与服务端版本不一致的问题

## Version 1.1.0
- 优化部分代码，使变得更可读
- 调整客户端端口映射配置，使变得更容易理解
- 修复客户端占用大量 CPU 的BUG（严重）
- 增加访问端口范围限制

## Version 1.2.0
- 修复由于客户端断开，未关闭监听造成启动后刷出大堆访问日志的BUG
- 重构服务端代码，调整服务端架构，解决服务端架构带来的`accept`问题
- 简化通讯协议，去除原端口参数
- 去除客户端参数“max-redial-times”，默认掉线无限重连
- 增加客户端参数“tunnel-count”，隧道条数，默认为1，范围\[1-5\]

## Version 1.2.1
- 优化代码，提升通讯效率

## Version 1.2.2
- 使用 go.mod 管理依赖
- 修复网络波动等因素会导致服务端不能正常提供服务的问题

## Version 1.3.0
- 通讯协议增加客户端ID标识，使用机器码
- 服务端增加访问端口校验，一个访问端口不能被重复占用
- 增加超时机制，每30分钟检测一遍隧道活性，失活则重建隧道

## Version 1.4.0
- 重构隧道活性检测机制，每1分钟检测隧道活性，容错率提高
- 修复端口占用不能准确报错的BUG


## TODO

- 服务端增加“最大端口数”配置，避免无限制开放端口
- 增加心跳机制检测服务是否通畅
- 通讯协议加密
- 增加命令行，查看端口列表
- 增加黑名单，支持屏蔽IP

通讯协议

- 通讯结果    1个字节(0: 成功，其他：失败)
- 版本号      4个字节
- 访问端口    4个字节
- 客户端ID    32位uuid
- Key        建议长度 6-16 字符串

协议最大长度不能超过 255

举例
1|1|13306|machineID|winshu

## 简易编译打包脚本

```shell script
# linux
git pull
go env -w GOPROXY=https://goproxy.cn,direct
go build -o netbus main.go
mkdir netbus_linux_amd64
mv netbus netbus_linux_amd64
cp config_demo.ini netbus_linux_amd64/config.ini
tar -zvcf netbus_linux_amd64.tar.gz netbus_linux_amd64
rm -rf netbus_linux_amd64
mv netbus_linux_amd64.tar.gz /mnt/d/

```

```bash
# windows
git pull
go build -o netbus.exe main.go
mkdir netbus_windows_amd64
move /Y netbus.exe netbus_windows_amd64
copy /Y config_demo.ini netbus_windows_amd64\config.ini
xcopy /Y doc\*.bat netbus_windows_amd64
if exist netbus_windows_amd64.zip ( del /Q netbus_windows_amd64.zip )
"C:/Program Files (x86)/WinRAR/WinRAR.exe" a netbus_windows_amd64.zip netbus_windows_amd64

```




test/ip_test.go
package test

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"math"
	"os"
	"strconv"
	"strings"
	"testing"
)

type CnIp struct {
	value int
	max   int
}

func (p *CnIp) Contains(ip int) bool {
	return ip >= p.value && ip <= p.value+p.max
}

func (p *CnIp) Format() string {
	return fmt.Sprintf("%s~%s", p.print(p.value), p.print(p.value+p.max))
}

func (p *CnIp) print(ip int) string {
	return fmt.Sprintf("%d.%d.%d.%d",
		ip>>24, (ip&0xFFFFFF)>>16, (ip&0xFFFF)>>8, ip&0xFF)
}

func IpToNumber(ip string) (result int, err error) {
	slices := strings.Split(ip, ".")
	length := len(slices)
	if length != 4 {
		return -1, errors.New("illegal ip string length")
	}

	for index, slice := range slices {
		i, _ := strconv.Atoi(slice)
		if i < 0 || i >= 256 {
			return -1, errors.New("illegal ip strings")
		}
		result += i * int(math.Pow(256, float64(length-index-1)))
	}
	return result, nil
}

var CnIpPool = make([]CnIp, 0)

func init() {
	fi, err := os.Open("../cn_ips.txt")
	if err != nil {
		fmt.Printf("Error: %s\n", err)
		return
	}
	defer fi.Close()
	br := bufio.NewReader(fi)
	for {
		a, _, c := br.ReadLine()
		if c == io.EOF {
			break
		}
		lines := strings.Split(string(a), ",")
		max, _ := strconv.Atoi(strings.TrimSpace(lines[1]))

		ip, _ := IpToNumber(lines[0])
		local := CnIp{ip, max}
		CnIpPool = append(CnIpPool, local)
	}
	fmt.Println("Init CnIp success. size", len(CnIpPool))
}

func IsCnIp(ip string) bool {
	if ipNumber, err := IpToNumber(ip); err == nil {
		for _, cnIp := range CnIpPool {
			if cnIp.Contains(ipNumber) {
				//fmt.Println(cnIp.Format())
				return true
			}
		}
	}
	return false
}

func TestIpToNumber(t *testing.T) {
	fmt.Println(IpToNumber("39.184.149.118"))
	fmt.Println(IpToNumber("39.128.0.0"))
}

func TestContainsIp(t *testing.T) {
	fmt.Println("--国内IP判断--")
	var ips = []string{"39.184.149.118", "121.17.142.60", "45.141.87.9"}

	for _, ip := range ips {
		fmt.Printf("%s %t\n", ip, IsCnIp(ip))
	}
}

func TestContainsIp2(t *testing.T) {
	var ip, _ = IpToNumber("39.184.149.118")
	var value, _ = IpToNumber("39.128.0.0")
	var cnIp = CnIp{value, 4194304}

	fmt.Println(cnIp.Contains(ip))
}


test/tcp_test.go
package test

import (
	"bytes"
	"encoding/binary"
	"github.com/denisbrodbeck/machineid"
	"log"
	"net"
	"testing"
)

// 测试连接中断

func TestTCPServer(t *testing.T) {
	listener, _ := net.Listen("tcp", ":6666")
	for {
		conn, err := listener.Accept()
		if err != nil {
			log.Println(err)
			continue
		}

		var length uint32

		if err = binary.Read(conn, binary.BigEndian, &length); err != nil {
			log.Println(err)
			_ = conn.Close()
			return
		}

		// 读取协议内容
		body := make([]byte, length)
		if err = binary.Read(conn, binary.BigEndian, &body); err != nil {
			log.Println(err)
			_ = conn.Close()
			return
		}
		log.Println(string(body))
		_ = conn.Close()
	}
}

func TestTCPClient(t *testing.T) {
	conn, err := net.Dial("tcp", ":6666")
	if err != nil {
		log.Println("dial failed")
		return
	}
	body, _ := machineid.ID()

	var buf = make([]byte, 4)
	binary.BigEndian.PutUint32(buf, uint32(len(body)))

	buffer := bytes.NewBuffer([]byte{})
	buffer.Write(buf)
	buffer.WriteString(body)

	if _, err := conn.Write(buffer.Bytes()); err != nil {
		log.Println(err)

		_ = conn.Close()
	}
}


test/web_test.go
package test

import (
	"fmt"
	"log"
	"net/http"
	"testing"
)

// 本地模拟多个 web 服务

func init() {
	log.SetFlags(log.Lshortfile | log.LstdFlags | log.Lmicroseconds)
}

func handleWebRequest(w http.ResponseWriter, r *http.Request) {
	log.Println("path", r.URL.Path)
	_, _ = fmt.Fprintf(w, r.Host)
}

func listenOnPort(port int) {
	log.Println("Listen to port", port)
	http.HandleFunc("/", handleWebRequest)
	err := http.ListenAndServe(fmt.Sprintf(":%d", port), nil) //设置监听的端口
	if err != nil {
		log.Fatal("ListenAndServe: ", err)
	}
}

func TestWeb7001(t *testing.T) {
	listenOnPort(7001)
}

func TestWeb7002(t *testing.T) {
	listenOnPort(7002)
}


core/client.go
package core

import (
	"chuantou/config"
	"github.com/denisbrodbeck/machineid"
	"log"
	"net"
	"os"
	"runtime"
	"strings"
)

// 客户端ID，启动时就确定了
var clientID string

func init() {
	if id, err := machineid.ID(); err == nil {
		clientID = strings.ReplaceAll(id, "-", "")
		log.Println("Get client machine ID :", clientID)
	} else {
		log.Println("Fail to get machine ID")
		os.Exit(0)
	}
}

// 处理客户端连接
func handleClientConnection(cfg config.ClientConfig, index int) {
	connChan := make(chan net.Conn)
	flagChan := make(chan bool)

	// 远程拨号，建桥
	go buildTunnelConnection(cfg, index, connChan, flagChan)
	// 本地连接拨号，并建立双向通道
	go buildLocalConnection(cfg.LocalAddr[index], connChan, flagChan)
	// 初始化连接
	for i := 0; i < cfg.TunnelCount; i++ {
		flagChan <- true
	}
	log.Printf("Initilization tunnel [%d] [%d]", cfg.LocalAddr[index].Port2, cfg.TunnelCount)
}

func buildTunnelConnection(cfg config.ClientConfig, index int, connCh chan net.Conn, flagCh chan bool) {
	server := cfg.ServerAddr
	local := cfg.LocalAddr[index]

	for {
		select {
		case <-flagCh:
			// 新建协程，向桥端建立连接
			go func(ch chan net.Conn) {
				conn := dial(server, maxRetryTimes)
				if conn == nil {
					runtime.Goexit()
				}

				request := Protocol{
					Result:  protocolResultSuccess,
					Version: Version,
					Port:    local.Port2,
					ID:      clientID,
					Key:     cfg.Key,
				}

				if !sendProtocol(conn, request) {
					log.Fatalln("Fail to start. exit")
				}
				var response Protocol
			loop:
				for {
					// 此处会阻塞，以等待访问者连接
					response = receiveProtocol(conn)

					// 处理连接结果
					switch response.Result {
					case protocolResultHeartBeat:
						// 不做任何处理，继续监听
						//log.Println("heartbeat...", response.Port)
					case protocolResultSuccess:
						log.Printf("New connection [%d] [%s]\n", local.Port2, local.String())
						ch <- conn

						// 跳出循环
						break loop
					case protocolResultVersionMismatch:
						// 版本不匹配，退出客户端
						// 鉴权失败，退出客户端
						log.Fatalln("Version mismatch. exit")
					case protocolResultFailToAuth:
						// 鉴权失败，退出客户端
						log.Fatalln("Fail to auth. exit")
					case protocolResultIllegalAccessPort:
						// 访问端口不合法
						log.Fatalln("Illegal Access Port. exit")
					case protocolResultPortIsOccupied:
						// 访问端口被占用
						log.Fatalf("Port[%d] is occupied\n", response.Port)
					case protocolResultFail:
						log.Fatalln("Fail to start. exit")
					case protocolResultFailToReceive:
						// 一般是超时导致，不打印日志
						closeConn(conn)
						flagCh <- true

						// 跳出循环
						break loop
					default:
						// 连接中断，重新连接
						log.Printf("Tunnel connection interrupted, try to redial. [result=%d] [%s]\n", response.Result, local.String())
						closeConn(conn)
						flagCh <- true

						// 跳出循环
						break loop
					}
				}
			}(connCh)
		}
	}
}

// 本地服务连接拨号，并建立双向通道
func buildLocalConnection(local config.NetAddress, connCh chan net.Conn, flagCh chan bool) {
	for {
		select {
		case cn := <-connCh:
			// 建立本地连接访问
			go func(conn net.Conn) {
				// 本地连接，不需要重新拨号
				if localConn := dial(local, 0); localConn != nil {
					// 通知创建新桥
					flagCh <- true
					forward(localConn, conn)
				} else {
					// 放弃连接，重新建桥
					closeConn(conn)
					flagCh <- true
				}
			}(cn)
		}
	}
}

// 入口
func Client(cfg config.ClientConfig) {
	log.Println("Load config", cfg)

	// 遍历所有端口
	for index := range cfg.LocalAddr {
		go handleClientConnection(cfg, index)
	}

	select {}
}


core/common.go
package core

import (
	"chuantou/config"
	"fmt"
	"io"
	"log"
	"net"
	"sync"
	"time"
)

const (
	// 重连间隔时间
	retryIntervalTime = 5

	// 心跳间隔时间
	heartBeatIntervalTime = 60 * time.Second

	// 最大重试次数
	maxRetryTimes = 24 * 60 * 60 / retryIntervalTime
)

var bufferPool *sync.Pool

func init() {
	log.Println("Init copy buffer pool...")
	bufferPool = &sync.Pool{}
	bufferPool.New = func() interface{} {
		return make([]byte, 32*1024)
	}
}

// 使用 bufferPool 重写 copy 函数， 避免反复 gc，提升性能
func copyWithPool(dst io.Writer, src io.Reader) (written int64, err error) {
	if wt, ok := src.(io.WriterTo); ok {
		return wt.WriteTo(dst)
	}
	if rt, ok := dst.(io.ReaderFrom); ok {
		return rt.ReadFrom(src)
	}

	buf := bufferPool.Get().([]byte)
	defer bufferPool.Put(buf)
	for {
		nr, er := src.Read(buf)
		if nr > 0 {
			nw, ew := dst.Write(buf[0:nr])
			if nw > 0 {
				written += int64(nw)
			}
			if ew != nil {
				err = ew
				break
			}
			if nr != nw {
				err = io.ErrShortWrite
				break
			}
		}
		if er == io.EOF {
			break
		}
		if er != nil {
			err = er
			break
		}
	}
	return written, err
}

// 连接数据复制
func connCopy(dist, source net.Conn, wg *sync.WaitGroup) {
	if _, err := copyWithPool(dist, source); err != nil {
		//log.Println("Connection interrupted", err)
	}
	_ = dist.Close()
	wg.Done()
}

// 连接转发
func forward(conn1, conn2 net.Conn) {
	//log.Printf("Forward channel [%s/%s] <-> [%s/%s]\n",
	//	conn1.RemoteAddr(), conn1.LocalAddr(), conn2.RemoteAddr(), conn2.LocalAddr())

	var wg sync.WaitGroup
	// wait tow goroutines
	wg.Add(2)
	go connCopy(conn1, conn2, &wg)
	go connCopy(conn2, conn1, &wg)
	//blocking when the wg is locked
	wg.Wait()
}

// 关闭连接
func closeConn(connections ...net.Conn) {
	for _, conn := range connections {
		if conn != nil {
			_ = conn.Close()
		}
	}
}

// 拨号
func dial(targetAddr config.NetAddress /*目标地址*/, maxRedialTimes int /*最大重拨次数*/) net.Conn {
	redialTimes := 0
	for {
		conn, err := net.Dial("tcp", targetAddr.String())
		if err == nil {
			//log.Printf("Dial to [%s] success.\n", targetAddr)
			return conn
		}
		redialTimes++
		if maxRedialTimes < 0 || redialTimes < maxRedialTimes {
			// 重连模式，每5秒一次
			log.Printf("Dial to [%s] failed, redial(%d) after %d seconeds.", targetAddr.String(), redialTimes, retryIntervalTime)
			time.Sleep(retryIntervalTime * time.Second)
		} else {
			log.Printf("Dial to [%s] failed. %s\n", targetAddr.String(), err.Error())
			return nil
		}
	}
}

// 监听端口
func listen(port uint32, id string) net.Listener {
	address := fmt.Sprintf("0.0.0.0:%d", port)
	listener, err := net.Listen("tcp", address)
	if err != nil {
		log.Println("Listen failed, the port may be used or closed", port)
		return nil
	}
	log.Printf("Listening at address %s by %s\n", address, id)
	return listener
}

// 受理请求
func accept(listener net.Listener) net.Conn {
	conn, err := listener.Accept()
	if err != nil {
		log.Println("Accept connect failed ->", err.Error())
		return nil
	}
	//log.Println("Accept a new client ->", conn.RemoteAddr())
	return conn
}

// 定时器，每间隔一段时间执行一遍函数
func setInterval(callback func(), duration time.Duration) {
	c := time.Tick(duration)
	for {
		callback()
		<-c
	}
}


core/server.go
package core

import (
	"chuantou/config"
	"log"
	"net"
	"sync"
	"time"
)

// 隧道连接
type TunnelConn struct {
	conn       net.Conn
	createTime time.Time
}

// 隧道上下文
type TunnelContext struct {
	request    Protocol        // 请求信息
	listener   net.Listener    // 服务端监听
	tunnelChan chan TunnelConn // 连接池
	createTime time.Time       // 创建时间
	lastTime   time.Time       // 最后检查时间
}

// 存放连接
func (p *TunnelContext) pushConn(conn net.Conn) {
	p.tunnelChan <- TunnelConn{
		conn:       conn,
		createTime: time.Now(),
	}
}

// 心跳，检测连接活性
// 连接池中有连接，则返回成功
func (p *TunnelContext) hearBeat() bool {
	tunnelCount := len(p.tunnelChan)

	for i := 0; i < tunnelCount; i++ {
		tunnelConn := <-p.tunnelChan
		//log.Println("heartbeat...", tunnelConn.createTime.UnixNano())

		// 检测活性
		if sendProtocol(tunnelConn.conn, p.request.NewResult(protocolResultHeartBeat)) {
			// 将连接重新放回连接池
			p.tunnelChan <- tunnelConn
		} else {
			// 关闭失去活性的连接
			closeConn(tunnelConn.conn)
		}
	}
	// 当连接池中没有可用连接时，心跳失败
	p.lastTime = time.Now()
	return len(p.tunnelChan) > 0
}

// key:   accessPort
// value: TunnelContext
var (
	tunnelContextMap   sync.Map
	tunnelContextMutex sync.Mutex
)

// 处理隧道连接
func handleTunnelConnection(tunnelConn net.Conn, cfg config.ServerConfig, tunnelContextChan chan TunnelContext) {
	// 接收协议消息
	req := receiveProtocol(tunnelConn)

	// 检查请求合法性
	if protocolResult := checkRequest(req, cfg); protocolResult != protocolResultSuccess {
		log.Printf("Illegal request, code = %b, ip = %s\n", protocolResult, tunnelConn.RemoteAddr().String())
		sendProtocol(tunnelConn, req.NewResult(protocolResult))
		closeConn(tunnelConn)
		return
	}

	// 获取隧道连接
	context, exists := tunnelContextMap.Load(req.Port)
	if exists {
		tunnelContext := context.(TunnelContext)
		// 端口的开启者是当前访问者
		if tunnelContext.request.IsSameID(&req) {
			tunnelContext.pushConn(tunnelConn)
		} else {
			// 端口已经被其他客户端占用，返回相应提示
			sendProtocol(tunnelConn, req.NewResult(protocolResultPortIsOccupied))
			closeConn(tunnelConn)
		}
		return
	}

	// 第一次创建才会执行，避免每次都加锁
	registerTunnelContext(req, tunnelContextChan)
	tunnelContextMutex.Lock()
	defer tunnelContextMutex.Unlock()
	context, exists = tunnelContextMap.Load(req.Port)
	if !exists {
		context = TunnelContext{
			request:    req,
			listener:   listen(req.Port, req.ID),
			tunnelChan: make(chan TunnelConn, config.MaxTunnelCount),
			createTime: time.Now(),
			lastTime:   time.Now(),
		}
		tunnelContextMap.Store(req.Port, context)
		tunnelContextChan <- context.(TunnelContext)

		log.Printf("Register port [%d] [%s] [%s]\n", req.Port, tunnelConn.RemoteAddr().String(), req.ID)
	}
	tunnelContext := context.(TunnelContext)
	tunnelContext.pushConn(tunnelConn)
}

func registerTunnelContext(req Protocol, tunnelContextChan chan TunnelContext) TunnelContext {
	tunnelContextMutex.Lock()
	defer tunnelContextMutex.Unlock()

	context, exists := tunnelContextMap.Load(req.Port)
	if !exists {
		context = TunnelContext{
			request:    req,
			listener:   listen(req.Port, req.ID),
			tunnelChan: make(chan TunnelConn, config.MaxTunnelCount),
			createTime: time.Now(),
			lastTime:   time.Now(),
		}
		tunnelContextMap.Store(req.Port, context)
		tunnelContextChan <- context.(TunnelContext)
	}
	return context.(TunnelContext)
}

// 检查请求信息，返回结果
func checkRequest(req Protocol, cfg config.ServerConfig) byte {
	if !req.Success() {
		return req.Result
	}
	// 检查版本号
	if req.Version != Version {
		log.Println("Version mismatch", req.String())
		return protocolResultVersionMismatch
	}
	// 检查权限
	if _, ok := config.CheckKey(cfg.Key, req.Key); !ok {
		log.Println("Unauthorized access", req.String())
		return protocolResultFailToAuth
	}
	// 检查访问端口是否在允许范围内
	if ok := cfg.PortInRange(req.Port); !ok {
		log.Println("Access Port out of range", req.String())
		return protocolResultIllegalAccessPort
	}
	return protocolResultSuccess
}

// 处理访问连接
func handleServerConnection(context TunnelContext) {
	serverListener := context.listener
	if serverListener == nil {
		tunnelContextMap.Delete(context.request.Port)
		return
	}
	for {
		serverConn := accept(serverListener)
		if serverConn == nil {
			// 受理监听失败，可能是监听关闭了，结束连接
			break
		}
		// 取隧道连接
		tunnelConn := <-context.tunnelChan
		if sendProtocol(tunnelConn.conn, context.request) {
			log.Printf("Accept connection [%d] [%s]\n", context.request.Port, serverConn.RemoteAddr().String())
			go forward(tunnelConn.conn, serverConn)
		} else {
			log.Printf("No tunnel available, close server listener. [%d]\n", context.request.Port)
			closeConn(serverConn)
			_ = serverListener.Close()
			tunnelContextMap.Delete(context.request.Port)
			break
		}
	}
}

// 入口
func Server(cfg config.ServerConfig) {
	log.Println("Load config", cfg)

	// 监听隧道端口
	tunnelListener := listen(cfg.Port, "server")
	if tunnelListener == nil {
		log.Fatalln("Fail to listen the tunnel port.")
	}

	tunnelContextChan := make(chan TunnelContext)
	// 处理来自客户端的隧道请求
	go func() {
		for {
			tunnelConn := accept(tunnelListener)
			if tunnelConn != nil {
				go handleTunnelConnection(tunnelConn, cfg, tunnelContextChan)
			}
		}
	}()

	// 处理监听及访问
	go func() {
		for {
			select {
			case context := <-tunnelContextChan:
				go handleServerConnection(context)
			}
		}
	}()

	// 心跳，需要考虑端口过多，心跳时间不够的情况
	go setInterval(func() {
		tunnelContextMap.Range(func(key, value interface{}) bool {
			tunnelContext := value.(TunnelContext)
			if !tunnelContext.hearBeat() {
				_ = tunnelContext.listener.Close()
				tunnelContextMap.Delete(key)
			}
			return true
		})
	}, heartBeatIntervalTime)

	select {}
}


core/protocol.go
package core

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"net"
	"time"
)

const (
	// 协议-结果
	protocolResultSuccess           = 0 // 成功，默认值
	protocolResultFail              = 1 // 失败
	protocolResultHeartBeat         = 2 // 心跳
	protocolResultFailToReceive     = 3 // 接收失败
	protocolResultFailToAuth        = 4 // 鉴权失败
	protocolResultVersionMismatch   = 5 // 版本不匹配
	protocolResultIllegalAccessPort = 6 // 访问端口不合法
	protocolResultPortIsOccupied    = 7 // 访问端口被占用

	// 协议发送超时时间
	protocolSendTimeout = 5 * time.Second
	// 协议接收超时时间
	protocolReceiveTimeout = heartBeatIntervalTime + 5*time.Second

	// 版本序列(单调递增)
	// 从右往左
	// 第1位为小版本号，用于修复BUG
	// 第2位为次版本号，用于增删功能
	// 第3位为主版本号，用于结构等大的升级
	Version = 140
)

// 协议格式
// 结果|版本号|访问端口|machineid|Key
// 1|2|13306|uuid|winshu

// 协议
type Protocol struct {
	Result  byte   // 结果：0 失败，1 成功
	Version uint32 // 版本号，单调递增
	Port    uint32 // 访问端口
	ID      string // 机器码
	Key     string // 身份验证
}

// 转字符串
func (p *Protocol) String() string {
	return fmt.Sprintf("%d|%d|%d|%s|%s", p.Result, p.Version, p.Port, p.ID, p.Key)
}

// 返回一个新结果
func (p *Protocol) NewResult(newResult byte) Protocol {
	return Protocol{
		Result:  newResult,
		Version: p.Version,
		Port:    p.Port,
		ID:      p.ID,
		Key:     p.Key,
	}
}

// 序列化
func (p *Protocol) Bytes() []byte {
	buffer := bytes.NewBuffer([]byte{})

	buffer.WriteByte(p.Result)
	_ = binary.Write(buffer, binary.BigEndian, p.Version)
	_ = binary.Write(buffer, binary.BigEndian, p.Port)
	buffer.WriteString(p.ID)
	buffer.WriteString(p.Key)
	return buffer.Bytes()
}

// 协议长度
func (p *Protocol) Len() byte {
	return byte(len(p.Bytes()))
}

// 是否成功
func (p *Protocol) Success() bool {
	return p.Result == protocolResultSuccess
}

// 检查是否是同一客户端
func (p *Protocol) IsSameID(other *Protocol) bool {
	return p.ID == other.ID
}

// 解析协议
func parseProtocol(body []byte) Protocol {
	// 检查 body 长度，是否合法
	if len(body) < 42 {
		return Protocol{Result: protocolResultFail}
	}
	return Protocol{
		Result:  body[0],
		Version: binary.BigEndian.Uint32(body[1:5]),
		Port:    binary.BigEndian.Uint32(body[5:9]),
		ID:      string(body[9:41]),
		Key:     string(body[41:]),
	}
}

// 发送协议
// 第一个字节为协议长度
// 协议长度只支持到255
func sendProtocol(conn net.Conn, req Protocol) bool {
	buffer := bytes.NewBuffer([]byte{})
	buffer.WriteByte(req.Len())
	buffer.Write(req.Bytes())

	// 设置写超时时间，避免连接断开的问题
	if err := conn.SetWriteDeadline(time.Now().Add(protocolSendTimeout)); err != nil {
		log.Println("Fail to set write deadline.", err.Error())
		return false
	}
	// 写协议内容
	if _, err := conn.Write(buffer.Bytes()); err != nil {
		log.Printf("Send protocol failed. [%s] %s\n", req.String(), err.Error())
		return false
	}
	// 清空写超时设置
	if err := conn.SetWriteDeadline(time.Time{}); err != nil {
		log.Println("Fail to clear write deadline.", err.Error())
		return false
	}
	//log.Println("Send protocol", req.String())
	return true
}

// 接收协议
// 第一个字节为协议长度
func receiveProtocol(conn net.Conn) Protocol {
	var err error
	var length byte

	// 设置读超时时间略大于心跳时间，避免连接断开的问题
	if err := conn.SetReadDeadline(time.Now().Add(protocolReceiveTimeout)); err != nil {
		log.Println("Fail to set read deadline.", err.Error())
		return Protocol{Result: protocolResultFailToReceive}
	}
	// 读取协议长度
	if err = binary.Read(conn, binary.BigEndian, &length); err != nil {
		return Protocol{Result: protocolResultFailToReceive}
	}
	// 读取协议内容
	body := make([]byte, length)
	if err = binary.Read(conn, binary.BigEndian, &body); err != nil {
		return Protocol{Result: protocolResultFailToReceive}
	}
	// 清空读超时设置
	if err := conn.SetReadDeadline(time.Time{}); err != nil {
		log.Println("Fail to clear read deadline.", err.Error())
		return Protocol{Result: protocolResultFailToReceive}
	}
	return parseProtocol(body)
}


main.go
package main

import (
	"chuantou/config"
	"chuantou/core"
	"fmt"
	"log"
	"os"
)

func init() {
	log.SetFlags(log.Ldate | log.Ltime | log.LstdFlags)

	fmt.Println("+----------------------------------------------------------------+")
	fmt.Println("|-------------------欢迎使用货云通网络代理系统----------------------|")
	fmt.Println("+----------------------------------------------------------------+")
	fmt.Println()
}

func printHelp() {
	fmt.Println(`A: "-server" load "config.ini" and start as server`)
	fmt.Println(`   "-client" load "config.ini" and start as client`)
	fmt.Println(`B: "-server <key> <port>" start as server, and listening at port x', e.g. -server 6666`)
	fmt.Println(`   "-client <key> <server:port> <local:port:mapping> [tunnel-count]" start as client,`)
	fmt.Println(`   "e.g. -client winshu 123.54.23.67:6666 127.0.0.1:3306:13306`)
	fmt.Println(`Generate trial key: `)
	fmt.Println(`   "-generate <key> [expired-time]" make a trial client key, e.g. -generate winshu 2019-12-31`)
	fmt.Println(`more details please read "README.md"`)
}

func main() {
	args := os.Args
	argc := len(os.Args)

	if argc < 2 {
		printHelp()
		os.Exit(0)
	}

	// 获取其余参数
	argsConfig := args[2:]

	switch args[1] {
	case "-server": //服务器端参数启动
		serverConfig := config.InitServerConfig(argsConfig)
		core.Server(serverConfig)
	case "-client": //客户端参数启动
		clientConfig := config.InitClientConfig(argsConfig)
		core.Client(clientConfig)
	case "-generate": //生成短期 key
		// 生成短期 key
		var seed, expired string
		if len(argsConfig) > 0 {
			seed = argsConfig[0]
		}
		if len(argsConfig) > 1 {
			expired = argsConfig[1]
		}
		if len(argsConfig) > 0 {
			trialKey, _ := config.NewKey(seed, expired)
			fmt.Println("You got a new key ->    ", trialKey)
		}
	case "-check":
		if len(argsConfig) == 2 {
			fmt.Println(config.CheckKey(argsConfig[0], argsConfig[1]))
		}
	case "-version":
		fmt.Println("Version", core.Version)
	default:
		printHelp()
	}
}
